File: Frontend/src/api/jira.js
import axios from "axios";

const API_URL = "http://localhost:8000"; // Backend URL

// Create an axios instance with default config
const api = axios.create({
  baseURL: API_URL,
});

// Add a request interceptor to add the auth token to all requests
api.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem("access_token");
    console.log("Token from localStorage:", token ? `${token.substring(0, 10)}...` : "No token found");
    
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
      console.log("Authorization header set:", `Bearer ${token.substring(0, 10)}...`);
    } else {
      console.warn("No token found in localStorage");
    }
    
    console.log("Request config:", {
      url: config.url,
      method: config.method,
      headers: config.headers,
      params: config.params
    });
    
    return config;
  },
  (error) => {
    console.error("Request interceptor error:", error);
    return Promise.reject(error);
  }
);

// Add a response interceptor to handle token expiration
api.interceptors.response.use(
  (response) => {
    console.log("Response received:", {
      status: response.status,
      data: response.data
    });
    return response;
  },
  async (error) => {
    console.error("Response error:", {
      status: error.response?.status,
      data: error.response?.data,
      message: error.message
    });
    
    const originalRequest = error.config;

    // If the error is due to an expired token and we haven't tried to refresh yet
    if (error.response?.status === 401 && !originalRequest._retry) {
      originalRequest._retry = true;

      try {
        // Try to refresh the token
        const refreshToken = localStorage.getItem("refresh_token");
        if (!refreshToken) {
          // No refresh token available, redirect to login
          console.warn("No refresh token available, redirecting to login");
          window.location.href = "/login";
          return Promise.reject(error);
        }

        console.log("Attempting to refresh token");
        const response = await axios.post(`${API_URL}/auth/refresh`, {
          refresh_token: refreshToken,
        });

        const { access_token, expires_in } = response.data;
        localStorage.setItem("access_token", access_token);
        if (expires_in) {
          localStorage.setItem("token_expires_at", Date.now() + expires_in * 1000);
        }

        console.log("Token refreshed successfully");
        // Retry the original request with the new token
        originalRequest.headers.Authorization = `Bearer ${access_token}`;
        return api(originalRequest);
      } catch (refreshError) {
        // If refresh fails, redirect to login
        console.error("Token refresh failed:", refreshError);
        window.location.href = "/login";
        return Promise.reject(refreshError);
      }
    }

    return Promise.reject(error);
  }
);

// Fetch all Jira issues
export const getJiraIssues = async (jql_query = "") => {
  console.log("Fetching Jira issues with query:", jql_query);
  try {
    const response = await api.get(`/dashboard?jql_query=${encodeURIComponent(jql_query)}`);
    console.log("Jira issues fetched successfully:", response.data);
    return response.data;
  } catch (error) {
    console.error("Error fetching Jira issues:", error);
    throw error;
  }
};

// Get all available transitions for a given issue key
export const getAvailableTransitions = async (issueKey) => {
  try {
    const response = await api.get(`/transitions/${issueKey}`);
    return response.data.transitions;
  } catch (error) {
    console.error("Error fetching transitions:", error);
    throw error;
  }
};

// Perform the issue transition by passing the issue key and transition ID
export const performTransition = async (issueKey, transitionId) => {
  try {
    const response = await api.post(`/transition/${issueKey}`, {
      transitionId: transitionId,
    });
    return response.data;
  } catch (error) {
    console.error("Error performing transition:", error);
    throw error;
  }
};

// Get a single Jira issue by key
export const getSingleIssue = async (issueKey) => {
  try {
    const response = await api.get(`/issues/${issueKey}`);
    return response.data;
  } catch (error) {
    console.error("Error fetching issue:", error);
    throw error;
  }
};

export const authenticateUser = async (code) => {
  try {
    const response = await api.get(`https://127.0.0.1:8443/auth/callback?code=${code}`, {
      headers: {
        'Accept': 'application/json'
      }
    });
    console.log("Authentication response:", response.data);
    return response.data;
  } catch (error) {
    console.error("Error during authentication:", error);
    throw error;
  }
};

File: Backend/app/routes/jira.py
from fastapi import APIRouter, Depends, HTTPException, Header
from app.services.jira_service import get_jira_issues, create_jira_ticket, get_jira_issue_by_key, transition_jira_issue, get_issue_transitions
from app.models import IssueCreate
# from app.utils.auth import get_current_user_token
from typing import Optional

router = APIRouter()

def get_token_from_header(authorization: Optional[str] = Header(None)):
    if not authorization or not authorization.startswith("Bearer "):
        raise HTTPException(status_code=401, detail="Not authenticated")
    return authorization.split(" ")[1]

@router.get("/dashboard")
async def dashboard(
    jql_query: str,
    authorization: Optional[str] = Header(None)
):
    """Retrieve Jira issues."""
    token = get_token_from_header(authorization)
    issues = get_jira_issues(jql_query=jql_query, access_token=token)
    if issues:
        return {"issues": issues}
    else:
        raise HTTPException(status_code=400, detail="No issues found")

@router.post("/create_ticket")
async def create_ticket(
    issue: IssueCreate,
    authorization: Optional[str] = Header(None)
):
    """Create a new Jira issue."""
    token = get_token_from_header(authorization)
    result = create_jira_ticket(issue, access_token=token)
    if result:
        return {"message": "Issue created successfully"}
    else:
        raise HTTPException(status_code=400, detail="Failed to create issue")

@router.get("/transitions/{issue_key}")
async def get_transitions(
    issue_key: str,
    authorization: Optional[str] = Header(None)
):
    """Get available transitions for a Jira issue."""
    token = get_token_from_header(authorization)
    transitions = get_issue_transitions(issue_key, access_token=token)
    if "error" in transitions:
        raise HTTPException(status_code=400, detail=transitions["error"])
    return {"transitions": transitions}

@router.post("/transition/{issue_key}")
async def transition_issue(
    issue_key: str,
    transitionId: dict,
    authorization: Optional[str] = Header(None)
):
    """Transition a Jira issue to a new status."""
    token = get_token_from_header(authorization)
    result = transition_jira_issue(issue_key, transitionId["transitionId"], access_token=token)
    if isinstance(result, dict) and "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])
    return {"message": "Issue transitioned successfully"}

@router.get("/issues/{issue_key}")
async def fetch_issue(
    issue_key: str,
    authorization: Optional[str] = Header(None)
):
    """Fetch a single Jira issue by key."""
    token = get_token_from_header(authorization)
    issue = get_jira_issue_by_key(issue_key, access_token=token)
    if issue:
        return issue
    raise HTTPException(status_code=404, detail="Issue not found")